ckanext/tracking/plugin.py

```diff
@@ -0,0 +1,112 @@
+import click
+import ckan.plugins as p
+
+from typing import Any, Callable
+
+from ckan.plugins import toolkit
+from ckan.types import Context, CKANApp
+from ckan.common import CKANConfig
+
+from .cli.tracking import tracking
+from .helpers import popular
+from .middleware import TrackingMiddleware
+from .model import TrackingSummary
+
+
+class TrackingPlugin(p.SingletonPlugin):
+    p.implements(p.IClick)
+    p.implements(p.IConfigurer)
+    p.implements(p.IMiddleware, inherit=True)
+    p.implements(p.IPackageController, inherit=True)
+    p.implements(p.ITemplateHelpers)
+
+    # IClick
+    def get_commands(self) -> "list[click.Command]":
+        return [tracking]
+
+    # IConfigurer
+    def update_config(self, config: CKANConfig) -> None:
+        toolkit.add_resource("assets", "tracking")
+        toolkit.add_template_directory(config, "templates")
+
+    # IMiddleware
+    def make_middleware(self, app: CKANApp, config: CKANConfig) -> Any:
+        app = TrackingMiddleware(app, config)
```

Comment:
In this way, you replace CkanFlaskApp with the TrackingMiddleware object for all plugins and the rest of the `make_flask_app` function. Imagine multiple plugins that register the middleware in the same way and the last middleware that uses `CKANApp.config` attribute:
```
def make_middleware(app, config):
  app.app.app.app.config ...
```

So instead of wrapping the app into a middleware, it's better to use [`after_request`](https://flask.palletsprojects.com/en/3.0.x/api/#flask.Flask.before_request) callback:

```
def make_middleware(app, config):
  app.after_request(function_that_does_something_for_tracking)
  return app
```

You'll need to adapt(and probably replace with a plain function) code of TrackingMidddleware, but it will keep the original value of the `app` and nothing will break in external extensions.

Reply:
Thanks! Agree with refactoring it since I have encountered so far several times the app.app.app... stuff.